provider/gandi/gandi.go

```diff
@@ -0,0 +1,268 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gandi
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/go-gandi/go-gandi"
+	"github.com/go-gandi/go-gandi/livedns"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+	"sigs.k8s.io/external-dns/provider"
+)
+
+const (
+	gandiCreate          = "CREATE"
+	gandiDelete          = "DELETE"
+	gandiUpdate          = "UPDATE"
+	gandiTTL             = 600
+	gandiLiveDNSProvider = "livedns"
+)
+
+type GandiChanges struct {
+	Action   string
+	ZoneName string
+	Record   livedns.DomainRecord
+}
+
+type GandiProvider struct {
+	provider.BaseProvider
+	LiveDNSClient LiveDNSClientAdapter
+	DomainClient  DomainClientAdapter
+	domainFilter  endpoint.DomainFilter
+	DryRun        bool
+}
+
+func NewGandiProvider(ctx context.Context, domainFilter endpoint.DomainFilter, dryRun bool) (*GandiProvider, error) {
+	key, ok := os.LookupEnv("GANDI_KEY")
+	if !ok {
+		return nil, errors.New("no environment variable GANDI_KEY provided")
+	}
+	sharingID, _ := os.LookupEnv("GANDI_SHARING_ID")
+
+	g := gandi.Config{
+		SharingID: sharingID,
+		Debug:     false,
+		// dry-run doesn't work but it won't hurt passing the flag
```

Comment:
this confuses me 😆 , why is the dry-run not working?

Reply:
Apparently it's only for payment related functions which we're not using at all. The parameter-name is a bit misleading...